substrate/frame/message-queue/src/tests.rs

```diff
@@ -2034,3 +2034,103 @@ fn process_message_ok_true_keeps_storage_changes() {
 		assert_eq!(sp_io::storage::exists(b"key"), true);
 	});
 }
+
+#[test]
+fn force_set_head_can_starve_other_queues() {
+	use MessageOrigin::*;
+	build_and_execute::<Test>(|| {
+		// Enqueue messages to three queues.
+		for _ in 0..2 {
+			MessageQueue::enqueue_message(msg("A"), Here);
+			MessageQueue::enqueue_message(msg("B"), There);
+			MessageQueue::enqueue_message(msg("C"), Everywhere(0));
+		}
+
+		// Servicing will only touch `Here` and `There`.
+		MessageQueue::service_queues(4.into_weight());
+		assert_eq!(
+			MessagesProcessed::take(),
+			vec![
+				(b"A".to_vec(), Here),
+				(b"A".to_vec(), Here),
+				(b"B".to_vec(), There),
+				(b"B".to_vec(), There)
```

Comment:
is it not A, B, C, A, because it process all messages of the first enqueued origin?

Reply:
It does not store the order of insertion, only the order of queues. So enqueing `A, B, C, A, B, C` becomes:  
Here: A A, There: B B, Everywhere: C C  
And it then starts at `Here`, processes all and then goes to `There`.

Reply:
That's the way I understood it. Thank you